### PR TITLE
feat: add agent address for validators

### PR DESCRIPTION
**Pull request by cosinlink**
### Description

add agent address for validators

### Rationale
Agent can execute `editConsensusAddress`, `editCommissionRate`, `editDescription`, `editVoteAddress` for his operator. 

### Changes

Notable changes:
* add agent address for validators
* ...

### CODE REVIEW

File: contracts/BC_fusion/StakeHub.sol [hunk 1]
```diff
@@ -315,6 +324,24 @@ contract StakeHub is System, Initializable, Protectable {
...
+    function updateAgent(address newAgent) external validatorExist(msg.sender) {
+        if (agentToOperator[newAgent] != address(0)) revert InvalidAgent();
+        if (_validatorSet.contains(newAgent)) revert InvalidAgent();
+
+        address operatorAddress = msg.sender;
+        address oldAgent = _validators[operatorAddress].agent;
+        if (oldAgent == newAgent) revert InvalidAgent();
+
+        if (oldAgent != address(0)) {
+            agentToOperator[oldAgent] = address(0);
```
Comment (pythonberg1997) [inline contracts/BC_fusion/StakeHub.sol:336]: why not use `delete agentToOperator[oldAgent]`

File: contracts/BC_fusion/StakeHub.sol [hunk 2]
```diff
@@ -332,6 +359,8 @@ contract StakeHub is System, Initializable, Protectable {
         // basic check
         address operatorAddress = msg.sender;
         if (_validatorSet.contains(operatorAddress)) revert ValidatorExisted();
+        if (agentToOperator[operatorAddress] != address(0)) revert InvalidAgent();
```
Comment (pythonberg1997) [inline contracts/BC_fusion/StakeHub.sol:362]: Is `InvalidAgent()` suitable here?

File: contracts/BC_fusion/StakeHub.sol [hunk 3]
```diff
@@ -1179,4 +1208,12 @@ contract StakeHub is System, Initializable, Protectable {
         uint256 bnbAmount = IStakeCredit(_validators[operatorAddress].creditContract).claim(msg.sender, requestNumber);
         emit Claimed(operatorAddress, msg.sender, bnbAmount);
     }
+
+    function _getMsgSender() internal view returns (address) {
```
Comment (pythonberg1997) [inline contracts/BC_fusion/StakeHub.sol:1212]: It may be better use `_getOperatorFromMsgSender`

Comment (SolidityGo) [reply contracts/BC_fusion/StakeHub.sol:1212]: `_getOperatorFromMsgSender ` may be inaccurate. In most cases, `msg.sender` is returned directly, and in rare cases, the `agent` can appear as a `msg.sender` replacement for the `operatorAddress`.

Comment (cosinlink) [reply contracts/BC_fusion/StakeHub.sol:1212]: `_getOperatorFromMsgSender ` may be inaccurate. In most cases, `msg.sender` is returned directly, and in rare cases, the `agent` can appear as a `msg.sender` replacement for the `operatorAddress`.

File: contracts/BC_fusion/StakeHub.sol (hunk 2 above)
Comment (cosinlink) [reply contracts/BC_fusion/StakeHub.sol]: fixed

File: contracts/BC_fusion/StakeHub.sol (hunk 3 above)
Comment (NathanBSC) [reply contracts/BC_fusion/StakeHub.sol:1214]: _getOperatorByMsgSender？

File: contracts/BC_fusion/StakeHub.sol (hunk 1 above)
Comment (NathanBSC) [reply contracts/BC_fusion/StakeHub.sol:338]: > why not use `delete agentToOperator[oldAgent]`

agree

File: contracts/BC_fusion/StakeHub.sol (hunk 3 above)
Comment (cosinlink) [reply contracts/BC_fusion/StakeHub.sol:1214]: This method is not to get `operator` address,  it's to get the real `msg.sender` for all address. Only in some scenarios, the agent can act as the operator's `msg.sender`.

Comment (cosinlink) [reply contracts/BC_fusion/StakeHub.sol:1214]: we can refer to the similar code here: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/05f218fb6617932e56bf5388c3b389c3028a7b73/contracts/metatx/ERC2771Context.sol#L55

File: contracts/BC_fusion/StakeHub.sol (hunk 1 above)
Comment (cosinlink) [reply contracts/BC_fusion/StakeHub.sol]: fixed

File: contracts/BC_fusion/StakeHub.sol [hunk 4]
```diff
@@ -332,6 +360,8 @@ contract StakeHub is System, Initializable, Protectable {
         // basic check
         address operatorAddress = msg.sender;
         if (_validatorSet.contains(operatorAddress)) revert ValidatorExisted();
+        if (agentToOperator[operatorAddress] != address(0)) revert InvalidValidator();
```
Comment (unclezoro) [inline contracts/BC_fusion/StakeHub.sol:363]: is this possible?

**Review by unclezoro — APPROVED** [review]

**Review by NathanBSC — APPROVED** [review]

**Review by MoharshBoru — COMMENTED** [review]
contracts/BC_fusion/StakeHub.sol

**Pull request merged**